src/trace-plugin-loader.ts

```diff
@@ -16,278 +16,485 @@
-                if (Object.keys(patchSet).length === 0) {
-                  logger.warn(`${moduleRoot}: version ${
-                      version} not supported by plugin.`);
+                  version = version.split('-')[0];
                 }
-              }
-            } else {
-              logger.warn(
-                  `${moduleRoot}: invalid version ${version}; not patching.`);
-            }
-            instrumentation.patches[moduleRoot] = patchSet;
-          }
 
-          for (const file of Object.keys(patchSet)) {
-            const patch = patchSet[file];
-            const loadPath =
-                moduleRoot ? path.join(moduleRoot, patch.file) : patch.file;
-            if (!patch.module) {
-              patch.module = originalModuleLoad(loadPath, module, false);
-            }
-            if (isInternalPatch(patch)) {
-              patch.patch(patch.module, instrumentation.agent);
-            }
-            if (isInternalIntercept(patch)) {
-              patch.module =
-                  patch.intercept(patch.module, instrumentation.agent);
-              intercepts[loadPath] = {interceptedValue: patch.module};
+                // Apply each supported plugin.
+                const plugin = this.pluginMap.get(name);
+                if (plugin) {
+                  if (plugin.isSupported(version)) {
+                    exportedValue =
+                        plugin.applyPlugin(exportedValue, file, version!);
+                  } else {
+                    this.logger.warn(`PluginLoader#onRequire: [${name}@${
+                        version}] This module is not supported by the configured set of plugins.`);
+                  }
+                }
+              } else if (isMainModule) {
+                this.logger.error(`PluginLoader#activate: [${
+                    name}] This module's version could not be determined. Not applying plugins.`);
+              }
             }
           }
-          const rootPatch = patchSet[''];
-          if (rootPatch && isInternalIntercept(rootPatch)) {
-            return rootPatch.module;
-          } else {
-            return null;
-          }
         }
+        return exportedValue;
+      });
+      this.internalState = PluginLoaderState.ACTIVATED;
+      this.logger.info(`PluginLoader#activate: Activated.`);
+    } else if (this.internalState === PluginLoaderState.DEACTIVATED) {
+      throw new Error('Currently cannot re-activate plugin loader.');
+    } else {
+      throw new Error('Plugin loader already activated.');
+    }
+    return this;
+  }
 
-        function moduleAlreadyPatched(
-            instrumentation: InternalPlugin, moduleRoot: string) {
-          return instrumentation.patches[moduleRoot];
-        }
+  /**
+   * Deactivates the plugin loader, preventing additional plugins from getting
+   * loaded or applied, as well as unpatching any modules for which plugins
+   * specified an unpatching method.
+   */
+  deactivate(): PluginLoader {
```

Comment:
`deactivate()` is idempotent while `activate()` is not. Would it be better to be consistent?